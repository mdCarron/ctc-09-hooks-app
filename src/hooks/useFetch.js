import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
  const [data, setData] = useState({ data: null, loading: true, error: null });

  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setData({ data: null, loading: true, error: false });

    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        if (isMounted.current) {
          setData({
            data,
            loading: false,
            error: false,
          });
        }
      });
  }, [url]);

  return data;
};
