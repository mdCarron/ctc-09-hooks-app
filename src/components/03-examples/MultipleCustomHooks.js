import React from "react";
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";

const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );
  console.log(counter);
  const { author, quote } = !!data && data[0];

  return (
    <div className="container mt-5">
      <h1>Breaking Bad Quotes</h1>
      <hr />

      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-0">{quote}</p>
          <footer className="blockquote-footer">{author}</footer>
        </blockquote>
      )}
      <button
        onClick={() => {
          increment(1);
        }}
        className="btn btn-primary"
      >
        Siguiente Frase
      </button>
    </div>
  );
};

export default MultipleCustomHooks;
