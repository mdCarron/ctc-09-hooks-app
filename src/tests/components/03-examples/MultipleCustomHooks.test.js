import React from "react";
import { shallow } from "enzyme";
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";

jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");

describe("<MultipleCustomHooks />", () => {
  useCounter.mockReturnValue({
    counter: 100,
    increment: () => {},
  });

  test("should match snapshot", () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should fetch and show information", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "Sopita",
          quote: "Hola Mundo",
        },
      ],
      loading: false,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper.find(".alert").exists()).toBe(false);
    expect(wrapper.find(".mb-0").text().trim()).toBe("Hola Mundo");
    expect(wrapper.find(".blockquote-footer").text().trim()).toBe("Sopita");
  });
});
