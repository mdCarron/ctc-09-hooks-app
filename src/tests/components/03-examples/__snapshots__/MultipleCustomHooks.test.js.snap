// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultipleCustomHooks /> should match snapshot 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente Frase
  </button>
</div>
`;
